### PR TITLE
[CALCITE-5180] Implement BigQuery TIME_TRUNC and TIMESTAMP_TRUNC functions

### PR DESCRIPTION
**Pull request by tjbanghart**

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/prepare/CalciteCatalogReader.java [hunk 1]
```diff
@@ -91,7 +91,10 @@ public class CalciteCatalogReader implements Prepare.CatalogReader {
 
   public CalciteCatalogReader(CalciteSchema rootSchema,
       List<String> defaultSchema, RelDataTypeFactory typeFactory, CalciteConnectionConfig config) {
-    this(rootSchema, SqlNameMatchers.withCaseSensitive(config != null && config.caseSensitive()),
+    this(rootSchema,
+        // TODO: BigQuery has case-insensitive functions
```
Comment (tjbanghart) [inline core/src/main/java/org/apache/calcite/prepare/CalciteCatalogReader.java:95]: https://github.com/apache/calcite/pull/2914 should fix this.

File: babel/src/test/resources/sql/big-query.iq [hunk 2]
```diff
@@ -16,7 +16,809 @@
 # limitations under the License.
 #
 !use scott-big-query
-!set outputformat csv
+!set outputformat mysql
```
Comment (julianhyde) [inline babel/src/test/resources/sql/big-query.iq:19]: when I merge this I'll factor out the big-query.iq change as a separate commit

File: testkit/src/main/java/org/apache/calcite/sql/parser/SqlParserTest.java [hunk 3]
```diff
@@ -520,7 +520,9 @@ public class SqlParserTest {
       "TEMPORARY",                     "92", "99",
       "THEN",                          "92", "99", "2003", "2011", "2014", "c",
       "TIME",                          "92", "99", "2003", "2011", "2014", "c",
+      "TIME_TRUNC", // BigQuery
```
Comment (julianhyde) [inline testkit/src/main/java/org/apache/calcite/sql/parser/SqlParserTest.java:523]: I'm surprised there are new keywords. We needed a new keyword for `ILIKE` because it's infix, but these can just be function names.

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java [hunk 4]
```diff
@@ -1604,6 +1604,10 @@ protected static Calendar getCalendarNotTooNear(int timeUnit) {
     f.checkScalar("{fn TIMESTAMPDIFF(MONTH,"
         + " TIMESTAMP '2019-09-01 00:00:00',"
         + " TIMESTAMP '2020-03-01 00:00:00')}", "6", "INTEGER NOT NULL");
+    f.checkScalar("{fn TIMESTAMP_TRUNC(TIMESTAMP '2019-09-20 15:30:00', MONTH) }",
```
Comment (julianhyde) [inline testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:1607]: I don't think JDBC syntax is a requirement. You should remove these lines, and revert the changes to `SqlJdbcFunctionCall.java`.

File: core/src/test/java/org/apache/calcite/test/SqlValidatorTest.java [hunk 5]
```diff
@@ -4083,6 +4083,55 @@ void subTestIntervalSecondNegative() {
...
+              + "TIME_TRUNC\\(<TIME>,.*\\)");
+      expr(String.format(Locale.ROOT, test, unit))
+          .withOperatorTable(opTable)
+          .columnType("TIME(0) NOT NULL");
+    }
+    // should fail on incompatible time unit
+    expr("TIME_TRUNC(TIME '15:30:00.00', ^DAY^)")
+        .fails("(?s).*Was expecting one of.*");
+  }
+
```
Comment (julianhyde) [inline core/src/test/java/org/apache/calcite/test/SqlValidatorTest.java:4134]: Are these tests superfluous? between `SqlParserTest` and `SqlOperatorTest` I think you have it covered. I don't think there's any non-trivial validation behavior.

File: site/_docs/reference.md [hunk 6]
```diff
@@ -2625,6 +2625,8 @@ semantics.
 | b m o p | SUBSTR(string, position [, substringLength ]) | Returns a portion of *string*, beginning at character *position*, *substringLength* characters long. SUBSTR calculates lengths using characters as defined by the input character set
 | m | STRCMP(string, string)                         | Returns 0 if both of the strings are same and returns -1 when the first argument is smaller than the second and 1 when the second one is smaller than the first one
 | o | TANH(numeric)                                  | Returns the hyperbolic tangent of *numeric*
+| b | TIME_TRUNC(time, timeUnit)                     | Truncates a *time* value to the granularity of *timeUnit*. The *time* value is always rounded to the beginning of timeUnit, which can be one of the following: MILLISECOND, SECOND, MINUTE, HOUR.
+| b | TIMESTAMP_TRUNC(timestamp, timeUnit)           | Truncates a *timestamp* value to the granularity of *timeUnit*. The *timestamp* value is always rounded to the beginning of the *timeUnit*.
```
Comment (julianhyde) [inline site/_docs/reference.md:2629]: move down a few lines, to retain alphabetical order

**Review by julianhyde — CHANGES_REQUESTED** [review]
almost there. one or two changes and I'll approve.

you don't need to do anything with big-query.iq - I'll handle that during merge.

File: core/src/test/java/org/apache/calcite/test/SqlValidatorTest.java (hunk 5 above)
Comment (tjbanghart) [reply core/src/test/java/org/apache/calcite/test/SqlValidatorTest.java:4134]: These can probably be simplified to run a single time unit but they validate for an error message if the BigQuery operator table is not used. I can add similar check in `SqlOperatorTest` so this file isn't touched.

File: testkit/src/main/java/org/apache/calcite/sql/parser/SqlParserTest.java (hunk 3 above)
Comment (tjbanghart) [reply testkit/src/main/java/org/apache/calcite/sql/parser/SqlParserTest.java:523]: This is me following an example commit too closely.
